src/cryptonote_core/blockchain.cpp

```diff
@@ -2571,248 +2548,200 @@ bool Blockchain::check_tx_inputs(transaction& tx, tx_verification_context &tvc,
+        continue;
+      }
+
+      if (!existing_tx.is_deregister_tx())
+        continue;
 
-#if defined(CACHE_VIN_RESULTS)
-      auto itk = it->second.find(in_to_key.k_image);
-      if(itk != it->second.end())
+      tx_extra_service_node_deregister existing_deregister;
+      if (!get_service_node_deregister_from_tx_extra(existing_tx.extra, existing_deregister))
       {
-        if(!itk->second)
-        {
-          MERROR_VER("Failed ring signature for tx " << get_transaction_hash(tx) << "  vin key with k_image: " << in_to_key.k_image << "  sig_index: " << sig_index);
-          return false;
-        }
+        MERROR_VER("could not get service node deregister from tx extra, possibly corrupt tx");
+        continue;
+      }
 
-        // txin has been verified already, skip
-        sig_index++;
+      const std::shared_ptr<const service_nodes::quorum_state> existing_deregister_quorum_state = m_service_node_list.get_quorum_state(existing_deregister.block_height);
+      if (!existing_deregister_quorum_state)
+      {
+        MERROR_VER("could not get quorum state for recent deregister tx");
         continue;
       }
-#endif
-    }
 
-    // make sure that output being spent matches up correctly with the
-    // signature spending it.
-    if (!check_tx_input(tx.version, in_to_key, tx_prefix_hash, tx.version == 1 ? tx.signatures[sig_index] : std::vector<crypto::signature>(), tx.rct_signatures, pubkeys[sig_index], pmax_used_block_height))
-    {
-      it->second[in_to_key.k_image] = false;
-      MERROR_VER("Failed to check ring signature for tx " << get_transaction_hash(tx) << "  vin key with k_image: " << in_to_key.k_image << "  sig_index: " << sig_index);
-      if (pmax_used_block_height) // a default value of NULL is used when called from Blockchain::handle_block_to_main_chain()
+      if (existing_deregister_quorum_state->nodes_to_test[existing_deregister.service_node_index] ==
+          quorum_state->nodes_to_test[deregister.service_node_index])
       {
-        MERROR_VER("  *pmax_used_block_height: " << *pmax_used_block_height);
+        MERROR_VER("Already seen this deregister tx (aka double spend)");
+        tvc.m_double_spend = true;
+        return false;
       }
+    }
+  }
+  else
+  {
+    crypto::hash tx_prefix_hash = get_transaction_prefix_hash(tx);
 
-      return false;
+    auto it = m_check_txin_table.find(tx_prefix_hash);
+    if(it == m_check_txin_table.end())
+    {
+      m_check_txin_table.emplace(tx_prefix_hash, std::unordered_map<crypto::key_image, bool>());
+      it = m_check_txin_table.find(tx_prefix_hash);
+      assert(it != m_check_txin_table.end());
     }
 
-    if (tx.version == 1)
+    std::vector<std::vector<rct::ctkey>> pubkeys(tx.vin.size());
+    std::vector < uint64_t > results;
+    results.resize(tx.vin.size(), 0);
+
+    size_t sig_index = 0;
+    const crypto::key_image *last_key_image = NULL;
+    for (const auto& txin : tx.vin)
```

Comment:
This should probably be turned into an old style for-loop, so we don't need (and don't forget) to increment sig_index manually.

Reply:
Fixed